### PR TITLE
Inductor: gate max-autotune 1x1 conv GEMM lowering

### LINKED ISSUE #161764: The "max-autotune" mode in torch.compile can lead to longer execution times relative to eager execution.
### 🐛 Describe the bug

I'm following the official demo to learn torch.compile, [doc](https://docs.pytorch.org/tutorials/intermediate/torch_compile_tutorial.html#best-practices-and-recommendations) torch.compile indeed save the time consuming when I set mode to reduce-overhead compared with eager in official demo, but when I changed the mode to max-autotune, it took double time than eager mode.
```python
# Returns the result of running `fn()` and the time it took for `fn()` to run,
# in seconds. We use CUDA events and synchronization for the most accurate
# measurements.
import torch
from torchsummary import summary

# 安装：pip install torchsummary


def timed(fn):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    result = fn()
    end.record()
    torch.cuda.synchronize()
    return result, start.elapsed_time(end) / 1000

# Generates random input and targets data for the model, where `b` is
# batch size.
def generate_data(b):
    return (
        torch.randn(b, 3, 128, 128).to(torch.float32).cuda(),
        torch.randint(1000, (b,)).cuda(),
    )

N_ITERS = 10

from torchvision.models import densenet121
def init_model():
    return densenet121().to(torch.float32).cuda()

model = init_model()

# Reset since we are using a different mode.
import torch._dynamo
torch._dynamo.reset()

# model_opt = torch.compile(model, mode="reduce-overhead")
model_opt = torch.compile(model, mode="max-autotune")
inp = generate_data(16)[0]
print(model)
with torch.no_grad():
    print("eager:", timed(lambda: model(inp))[1])
    model_opt(inp)
    print("compile:", timed(lambda: model_opt(inp))[1])
```

the output
```bash
eager: 0.3398114013671875
/home/zyl/miniconda3/envs/triton/lib/python3.11/site-packages/torch/_inductor/compile_fx.py:282: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
W0829 11:28:12.489000 187843 site-packages/torch/_inductor/utils.py:1436] [0/0] Not enough SMs to use max_autotune_gemm mode
AUTOTUNE addmm(16x1000, 16x1024, 1024x1000)
strides: [0, 1], [1024, 1], [1, 1024]
dtypes: torch.float32, torch.float32, torch.float32
  bias_addmm 0.0266 ms 100.0% 
  addmm 0.0299 ms 89.2% 
SingleProcess AUTOTUNE benchmarking takes 0.0584 seconds and 0.0002 seconds precompiling for 2 choices
compile: 0.6538045043945313
```

### Versions

Collecting environment information...
PyTorch version: 2.9.0a0+git323a690
Is debug build: True
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 14.2.0-4ubuntu2~24.04) 14.2.0
Clang version: Could not collect
CMake version: version 4.0.3
Libc version: glibc-2.39

Python version: 3.10.18 (main, Jun  5 2025, 13:14:17) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.15.9-061509-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.0.48
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 5060 Ti
Nvidia driver version: 580.65.06
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  20
On-line CPU(s) list:                     0-19
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) Ultra 7 265K
CPU family:                              6
Model:                                   198
Thread(s) per core:                      1
Core(s) per socket:                      20
Socket(s):                               1
Stepping:                                2
CPU(s) scaling MHz:                      26%
CPU max MHz:                             4600.0000
CPU min MHz:                             800.0000
BogoMIPS:                                7756.80
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni lam wbnoinvd dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid bus_lock_detect movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               704 KiB (18 instances)
L1i cache:                               1.1 MiB (18 instances)
L2 cache:                                36 MiB (11 instances)
L3 cache:                                30 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-19
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS Not affected; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] flake8==7.3.0
[pip3] flake8-bugbear==24.12.12
[pip3] flake8-comprehensions==3.16.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==2024.24.12
[pip3] flake8-pyi==25.5.0
[pip3] flake8_simplify==0.22.0
[pip3] mypy==1.16.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] optree==0.13.0
[pip3] torch==2.9.0a0+git323a690
[pip3] triton==3.4.0
[conda] numpy                      1.26.4              pypi_0           pypi
[conda] optree                     0.13.0              pypi_0           pypi
[conda] torch                      2.9.0a0+git323a690  dev_0            <develop>
[conda] torchfix                   0.4.0               pypi_0           pypi
[conda] triton                     3.4.0               pypi_0           pypi

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185159

The max-autotune 1x1 convolution lowering converted channels-last
convolutions to GEMM whenever max_autotune or max_autotune_gemm was
enabled. On GPUs that fail the GEMM template availability gate, this
could still replace a cudnn convolution with ATen mm/addmm. DenseNet has
many biased 1x1 convolutions, so this could make mode="max-autotune"
slower than eager after the warning that there are not enough SMs for
max_autotune_gemm.

Gate the automatic rewrite on a usable non-ATen template for the actual
converted op. Biased 1x1 convolutions require addmm-capable templates,
while unbiased 1x1 convolutions use the mm-capable template set. Keep the
explicit conv_1x1_as_mm knob unchanged, and only suppress bias peeling
when a layout-eligible biased 1x1 convolution would otherwise re-enter as
an unbiased mm plus pointwise add after addmm templates were rejected.

I kept this in the convolution lowering instead of changing compile mode
options or the global small-GPU gate because the bad behavior is the
local rewrite choosing a GEMM form without a usable GEMM template; other
max-autotune behavior can still be valid.

Fixes #161764
Generated by my agent

Test Plan:
- CUDA_VISIBLE_DEVICES=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_161764_manager_tests python test/inductor/test_max_autotune.py -k autotune_conv1x1
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto trunk 455c8139 and resubmitted it with the same code changes. The earlier ROCm failures were in AOT Triton autotuning/precomputed-grid and pad_mm tests outside this conv-lowering diff; local CUDA conv coverage, PadMMTest.test_exclude_padding, and lintrunner pass, and fresh CI is running.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main e0d9cf97915 and resubmitted it. The completed CI on the old head had ROCm failures while rerunning disabled issue 161764 tests and a cpp-wrapper failure in a Python-wrapper codegen assertion; current main removes/updates the stale failures, and I added cpp-wrapper skips for those Python-wrapper-only assertions. Local validation passes for normal and cpp-wrapper autotune_conv1x1 coverage plus lintrunner; fresh CI is now pending.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main 614c7affb27 and resubmitted it. The previous docs upload failures were environment errors with aws missing, and the TorchTitan jobs exited before running tests. The H100 failure is in an existing TMA addmm test and the local B200 run skips that SM90 path; the PR's conv1x1 max-autotune coverage passes locally. lintrunner currently reports unrelated Pyrefly errors in torch/export/pt2_archive/_package.py and torch/nn/modules/loss.py. Fresh CI is running on the rebased head.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resubmitted it at bd0a9273bbc. The prior Dr. CI new failure was an eager HuggingFace DebertaV2 import/TorchScript failure before compile, unrelated to this conv lowering. I also addressed a review-agent blocker by checking concrete CUTLASS GEMM choices before rewriting 1x1 convs to GEMM, so CUTLASS allowlist/denylist filters cannot leave only ATen addmm/mm choices. Local validation passes for the focused autotune_conv1x1 suite and lintrunner; fresh CI is running.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resubmitted it at 5d33209dfc8. The previous Dr. CI failures do not appear caused by this conv lowering change: B200 failed before test logic with CUDA devices unavailable, H100 hit the existing persistent TMA addmm yindex failure, and TorchTitan failed resolving distributed process groups in a torchcomms compile integration. The focused autotune_conv1x1 suite and lintrunner pass locally; fresh CI is now running.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resubmitted it at 2996ed92ead. The previous Dr. CI failures do not appear caused by this conv lowering change: B200 stopped abruptly around a conv2d backward smoke case outside the forward 1x1 rewrite path and then failed monitor cleanup, H100 hit the existing persistent TMA addmm yindex failure, and TorchTitan failed resolving torchcomms process groups. The focused autotune_conv1x1 suite and lintrunner pass locally; fresh CI is now running.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
